fms_extras/utils/generation.py

```diff
@@ -3,9 +3,623 @@
+    else:
+        next_vals = next_vals.view(
+            batch_size, n_candidates, decode_seq_length
+        )  # b k 1+h
+        embeds = embeds.view(
+            batch_size, n_candidates, decode_seq_length, embeds.size(2)
+        )  # b k 1+h d
+    return next_vals, embeds
+
+
+def speculative_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids: Union[torch.Tensor, List[torch.Tensor]],
+    speculator: MLPSpeculator,
+    kv_cache_manager: PagedKVCacheManager,
+    max_seq_len: int = 2048,
+    new_tokens: int = 256,
+    n_candidates: int = 5,
+    threshes=[5, 3, 2],
+    flattening: bool = True,
+    decode_model: Optional[Union[Callable, torch.nn.Module]] = None,
+    # todo: This is a WIP to enable cudagraphs, currently its only for batch_size=1
+    cudagraphs: bool = False,
+):
+    """
+    A reference implementation of speculative decoding generation.
+    Returns at least the specified number of tokens - the speculator may return a
+    few extra in the final step.
+    If input is batched, continues generating until EVERY sequence has produced AT LEAST the required number of tokens.
+    Input (and output) tokens beyond max_seq_len are simply dropped for a sliding-window approach.
+    Currently reproduces behavior of greedy decoding only.
+    Args:
+        model: A function or nn.Module that takes a batch of input_ids and
+            returns logits
+        input_ids: A length n tensor of token IDs, or list of such tensors
+        speculator: A function or nn.Module that takes a state vector and sampled token
+            and returns a set of candidate suffixes
+        kv_cache_manager: PagedKVCacheManager
+            the paged kv-cache manager to be used in generation
+        max_seq_len: the sequence length of the base model
+        new_tokens: number of tokens to generate
+        n_candidates: only consider the top n most confident candidates from the speculator
+        threshes: build candidate suffix trees by taking top-(threshes[n]) most confident values
+            for head n. len(threshes) must equal speculator.n_predict. prod(threshes) must be greater
+            than or equal to n_candidates (we cannot have more candidates than tree leaves).
+        flattening: enable batch flattening / tree attention with redundant prefix removal when compression
+            ratio is favorable. Adds extra overhead to shrink the token count in each batch.
+        decode_model: nn.Module, optional
+            an optional model that performs the decode forward step. If None, will use the function or nn.Module
+            provided to model param to perform decode forward. This parameter is intended to be used when the compile
+            flags for the model doing prefill and the model generating subsequent tokens are different (default is None)
+        cudagraphs: bool
+            if True, cudagraphs is used and all metadata will be padded, otherwise metadata will not be padded unless
```

Comment:
nit: it would be nice to wrap long lines at a shorter length, short enough that it's readable in a diff on github on a macbook without getting too garbled.

Reply:
fixed in latest commit